File: backend/apps/loans/models/loan_product.py
from django.db import models
from django.conf import settings
from apps.common.models import AuditModel


class InterestMethod(models.TextChoices):
    FLAT = "flat", "Flat"
    REDUCING_BALANCE = "reducing_balance", "Reducing Balance"
    COMPOUND = "compound", "Compound"


class InterestPeriod(models.TextChoices):
    MONTHLY = "monthly", "Monthly"
    YEARLY = "yearly", "Yearly"


class RepaymentFrequency(models.TextChoices):
    DAILY = "daily", "Daily"
    WEEKLY = "weekly", "Weekly"
    MONTHLY = "monthly", "Monthly"
    YEARLY = "yearly", "Yearly"


class SecurityType(models.TextChoices):
    PERCENTAGE = "percentage", "Percentage"
    FIXED_AMOUNT = "fixed_amount", "Fixed Amount"
    CUSTOM_TEXT = "custom_text", "Custom Text"


class FeeType(models.TextChoices):
    PERCENTAGE = "percentage", "Percentage"
    FIXED_AMOUNT = "fixed_amount", "Fixed Amount"


class DepositType(models.TextChoices):
    PERCENTAGE = "percentage", "Percentage"
    FIXED_AMOUNT = "fixed_amount", "Fixed Amount"


class LatePaymentPenaltyType(models.TextChoices):
    PERCENTAGE = "percentage", "Percentage"
    FIXED_AMOUNT = "fixed_amount", "Fixed Amount"


class OffsetCoverType(models.TextChoices):
    SAVINGS = "savings", "Savings"
    SECURITY = "security", "Security"
    BOTH = "both", "Both"


class LoanProduct(AuditModel):
    """
    Configurable loan product with immutable versioning.
    When a product is edited, a new version is spawned and the previous
    version is marked inactive, preserving existing loan contracts.
    """
    product_code = models.CharField(max_length=50, db_index=True)
    version_number = models.PositiveIntegerField(default=1)
    product_name = models.CharField(max_length=255)
    is_active = models.BooleanField(default=True, db_index=True)
    effective_date = models.DateField()

    # Interest Settings
    interest_method = models.CharField(
        max_length=30,
        choices=InterestMethod.choices,
        default=InterestMethod.REDUCING_BALANCE,
    )
    interest_rate = models.DecimalField(max_digits=8, decimal_places=4)
    interest_period = models.CharField(
        max_length=20,
        choices=InterestPeriod.choices,
        default=InterestPeriod.MONTHLY,
    )

    # Repayment Terms
    repayment_frequency = models.CharField(
        max_length=20,
        choices=RepaymentFrequency.choices,
        default=RepaymentFrequency.MONTHLY,
    )
    min_repayment_period = models.PositiveIntegerField(default=1)
    max_repayment_period = models.PositiveIntegerField(null=True, blank=True)
    min_amount = models.DecimalField(max_digits=15, decimal_places=2, default=0)
    max_amount = models.DecimalField(max_digits=15, decimal_places=2, null=True, blank=True)

    # Guarantor & Multipliers
    requires_guarantor = models.BooleanField(default=False)
    min_guarantors = models.PositiveIntegerField(default=0)
    is_multiple_of_savings = models.BooleanField(default=False)
    savings_multiplier = models.DecimalField(max_digits=8, decimal_places=4, null=True, blank=True)

    # Security / Collateral
    requires_security = models.BooleanField(default=False)
    security_type = models.CharField(
        max_length=30,
        choices=SecurityType.choices,
        null=True,
        blank=True,
    )
    security_value = models.DecimalField(max_digits=15, decimal_places=2, null=True, blank=True)
    security_notes = models.TextField(blank=True, null=True)

    # Deposit
    requires_deposit = models.BooleanField(default=False)
    deposit_type = models.CharField(
        max_length=30,
        choices=DepositType.choices,
        null=True,
        blank=True,
    )
    deposit_value = models.DecimalField(max_digits=15, decimal_places=2, null=True, blank=True)

    # Repayment Allocation Waterfall
    allocation_order = models.CharField(
        max_length=100,
        default="penalty,fees,interest,principal",
        help_text="Comma-separated payment allocation priority",
    )

    # Legacy / Default Penalty settings
    late_payment_penalty_type = models.CharField(
        max_length=30,
        choices=LatePaymentPenaltyType.choices,
        null=True,
        blank=True,
    )
    late_payment_penalty_value = models.DecimalField(
        max_digits=15, decimal_places=2, null=True, blank=True
    )
    grace_period_days = models.PositiveIntegerField(default=0)

    # Workflow controls
    requires_appraisal = models.BooleanField(default=False)
    requires_board_approval = models.BooleanField(default=False)

    # Aging & Delinquency Thresholds (days overdue)
    watchful_after_days = models.PositiveIntegerField(default=30)
    non_performing_after_days = models.PositiveIntegerField(default=90)
    doubtful_after_days = models.PositiveIntegerField(default=180)

    # Rescheduling & Offsets
    allows_rescheduling = models.BooleanField(default=False)
    reschedule_fee_type = models.CharField(
        max_length=30, choices=FeeType.choices, null=True, blank=True
    )
    reschedule_fee_value = models.DecimalField(
        max_digits=15, decimal_places=2, null=True, blank=True
    )

    allows_offset = models.BooleanField(default=False)
    offset_covers = models.CharField(
        max_length=30, choices=OffsetCoverType.choices, null=True, blank=True
    )
    offset_fee_type = models.CharField(
        max_length=30, choices=FeeType.choices, null=True, blank=True
    )
    offset_fee_value = models.DecimalField(
        max_digits=15, decimal_places=2, null=True, blank=True
    )

    organization = models.ForeignKey(
        "organizations.Organization",
        on_delete=models.PROTECT,
        related_name="loan_products",
        null=True,
        blank=True,
    )

    class Meta:
        db_table = "tbl_loan_products"
        ordering = ["product_code", "-version_number"]
        unique_together = [("product_code", "version_number")]

    def __str__(self):
        return f"{self.product_name} ({self.product_code} v{self.version_number})"


class LoanProductFee(models.Model):
    product = models.ForeignKey(
        LoanProduct, on_delete=models.CASCADE, related_name="fees"
    )
    fee_name = models.CharField(max_length=255)
    fee_type = models.CharField(max_length=30, choices=FeeType.choices)
    fee_value = models.DecimalField(max_digits=15, decimal_places=4)
    fee_basis = models.CharField(
        max_length=50,
        default="principal",
        help_text="principal | savings | deposit | loan_balance",
    )
    affects_principal = models.BooleanField(
        default=False,
        help_text="If true, fee amount is added to principal instead of deducted from disbursement",
    )
    show_in_statement = models.BooleanField(default=True)
    ledger_account_name = models.CharField(
        max_length=255, default="Loan Processing Fee Income"
    )

    class Meta:
        db_table = "tbl_loan_product_fees"

    def __str__(self):
        return f"{self.fee_name} - {self.fee_type} ({self.fee_value})"


class LoanProductPenalty(models.Model):
    product = models.ForeignKey(
        LoanProduct, on_delete=models.CASCADE, related_name="penalties"
    )
    penalty_name = models.CharField(max_length=255)
    penalty_type = models.CharField(
        max_length=30, choices=LatePaymentPenaltyType.choices
    )
    penalty_value = models.DecimalField(max_digits=15, decimal_places=2)
    grace_period_days = models.PositiveIntegerField(default=0)
    ledger_account_name = models.CharField(
        max_length=255, default="Penalty Income"
    )

    class Meta:
        db_table = "tbl_loan_product_penalties"

    def __str__(self):
        return f"{self.penalty_name} - {self.penalty_type} ({self.penalty_value})"

File: frontend/package/src/interfaces/loanProduct.ts
/**
 * Loan Product Interfaces
 * --------------------------------------------------------
 * Mirrors the Royal Loan Engine API exactly.
 */

import {
  LoanProductFee,
  LoanProductFeeCreate,
} from "./loanFee";

import {
  LoanProductPenalty,
  LoanProductPenaltyCreate,
} from "./loanPenalty";

/* ===========================
   ENUM TYPES
=========================== */

export type InterestMethod =
  | "flat"
  | "reducing_balance";

export type InterestPeriod =
  | "daily"
  | "weekly"
  | "monthly"
  | "yearly";

export type RepaymentFrequency =
  | "daily"
  | "weekly"
  | "monthly"
  | "quarterly"
  | "yearly";

export type PercentageOrFixed =
  | "percentage"
  | "fixed"
  | "fixed_amount";

export type OffsetCovers =
  | "savings"
  | "shares"
  | "deposit"
  | "security"
  | "both";

export type AllocationOrder =
  | "penalty,fees,interest,principal"
  | "penalty,interest,principal"
  | "interest,fees,penalty,principal"
  | "interest,penalty,principal"
  | "interest,principal,penalty"
  | "principal,interest,penalty"
  | string;

/* ===========================
   CREATE
=========================== */

export interface LoanProductCreate {
  product_code: string;

  product_name: string;

  effective_date: string;

  interest_method: InterestMethod;

  interest_rate: number;

  interest_period: InterestPeriod;

  repayment_frequency: RepaymentFrequency;

  max_repayment_period: number;

  requires_guarantor: boolean;

  is_multiple_of_savings: boolean;

  savings_multiplier: number;

  requires_security: boolean;

  security_type: PercentageOrFixed;

  security_value: number;

  security_notes: string;

  requires_deposit: boolean;

  deposit_type: PercentageOrFixed;

  deposit_value: number;

  late_payment_penalty_type: PercentageOrFixed;

  late_payment_penalty_value: number;

  requires_appraisal: boolean;

  requires_board_approval: boolean;

  watchful_after_days: number;

  non_performing_after_days: number;

  doubtful_after_days: number;

  allows_rescheduling: boolean;

  reschedule_fee_type: PercentageOrFixed;

  reschedule_fee_value: number;

  allows_offset: boolean;

  offset_covers: OffsetCovers;

  offset_fee_type: PercentageOrFixed;

  offset_fee_value: number;

  allocation_order: AllocationOrder;

  fees: LoanProductFeeCreate[];

  penalties: LoanProductPenaltyCreate[];
}

/* ===========================
   RESPONSE
=========================== */

export interface LoanProduct {
  id: number;

  product_code: string;

  version_number: number;

  product_name: string;

  is_active: boolean;

  effective_date: string;

  interest_method: InterestMethod;

  interest_rate: string;

  interest_period: InterestPeriod;

  repayment_frequency: RepaymentFrequency;

  min_repayment_period?: number;

  max_repayment_period: number | null;

  min_amount?: string | number;

  max_amount?: string | number | null;

  requires_guarantor: boolean;

  is_multiple_of_savings: boolean;

  savings_multiplier: string | null;

  requires_security: boolean;

  security_type: PercentageOrFixed | null;

  security_value: string | null;

  security_notes: string | null;

  requires_deposit: boolean;

  deposit_type: PercentageOrFixed | null;

  deposit_value: string | null;

  late_payment_penalty_type: PercentageOrFixed | null;

  late_payment_penalty_value: string | null;

  requires_appraisal: boolean;

  requires_board_approval: boolean;

  watchful_after_days: number | null;

  non_performing_after_days: number | null;

  doubtful_after_days: number | null;

  allows_rescheduling: boolean;

  reschedule_fee_type: PercentageOrFixed | null;

  reschedule_fee_value: string | null;

  allows_offset: boolean;

  offset_covers: OffsetCovers | null;

  offset_fee_type: PercentageOrFixed | null;

  offset_fee_value: string | null;

  allocation_order: AllocationOrder;

  created_at: string;

  updated_at: string;

  fees: LoanProductFee[];

  penalties: LoanProductPenalty[];
}

/* ===========================
   UPDATE
=========================== */

export interface LoanProductUpdate
  extends Partial<LoanProductCreate> {}

/* ===========================
   LIST
=========================== */

export type LoanProductList =
  LoanProduct[];